### PR TITLE
GCC7 Implicit-fallthrough

### PR DESCRIPTION
**Pull request by daveol**
This is the initial macro for solving switch case fallthrough errors. It
is non-empty in all major GCC versions higher than 7 and defines
```OLA_FALLTHROUGH``` as ```__attribute__ ((fallthrough));```

```OLA_FALLTHROUGH``` has to be used at the end of the case to work.

This fixes github bug #1193 and debian bug [853583](https://bugs.debian.org/853583)

### CODE REVIEW

File: include/ola/base/Macro.h [hunk 1]
```diff
@@ -63,6 +63,30 @@
...
+ *     case 'foo':
+ *       doFoo();
+ *       //after foo falltrough to bar because $REASON
+ *       OLA_FALLTROUGH
+ *     case 'bar':
+ *   @endcode
+ */
+#ifdef __GNUC__
+#if __GNUC__ >= 7
+#define OLA_FALLTROUGH __attribute__ ((fallthrough));
```
Comment (peternewman) [inline include/ola/base/Macro.h]: From the issue, we want to use [[fallthrough]]; when the compiler has C++17 support, regardless of which compiler it is, but that may be a future change I imagine. Perhaps add it as a todo if it doesn't work now (which it shouldn't as we're forcing something older to bypass other issues.

**Comment by peternewman** [comment]
This probably also wants to target 0.10 branch given it's fixing an existing bug.

File: include/ola/base/Macro.h (hunk 1 above)
Comment (daveol) [inline include/ola/base/Macro.h]: we compile with c++98 still.... so ```[[falltrough]];``` is not gonna work

**Comment by daveol** [comment]
I don't really like targeting branches that the bug is not specific for, #1168 demonstrates this perfectly because I now still need to apply the patch to every master every time I want to test.

Besides, you can still cherry-pick this into the 0.10 branch

File: include/ola/base/Macro.h (hunk 1 above)
Comment (peternewman) [inline include/ola/base/Macro.h]: True. Perhaps just add it as a future todo/comment for why we aren't using the [[]] variant.

**Comment by peternewman** [comment]
The first bit is easy enough to fix. As soon as this gets approved. https://github.com/OpenLightingProject/ola/pull/1205

I'd rather they go into 0.10 and flow nicely into master, which will be 0.10+more, than try and cherry-pick a change related to master that might relate to code that doesn't even exist in 0.10. Also the cherry-pick becomes a pain if there are multiple commits as there are with this. It's true we could squash, but personally I like to keep the history so things like the PR comments make sense.

File: plugins/usbpro/BaseRobeWidget.cpp [hunk 2]
```diff
@@ -118,16 +119,25 @@ void BaseRobeWidget::ReceiveMessage() {
           return;
       } while (m_header.som != SOM);
       m_state = RECV_PACKET_TYPE;
+
+      // if we don't return fallthrough
```
Comment (peternewman) [inline plugins/usbpro/BaseRobeWidget.cpp]: Pedant: "if we don't return, fallthrough"

File: include/ola/base/Macro.h [hunk 3]
```diff
@@ -63,6 +63,30 @@
...
 /**
+ * @def OLA_FALLTHROUGH
+ * @brief Mark switch cases as fallthrough when required
+ *
+ * @examplepara
+ *   @code
+ *   switch (cond) {
+ *     case 'foo':
+ *       doFoo();
+ *       //after foo fallthrough to bar because $REASON
```
Comment (peternewman) [inline include/ola/base/Macro.h]: Can you fix the lint of the example by adding a space between // and after please.

File: include/ola/base/Macro.h [hunk 4]
```diff
@@ -63,6 +63,30 @@
...
+ *       OLA_FALLTHROUGH
+ *     case 'bar':
+ *   @endcode
+ */
+#ifdef __GNUC__
+#if __GNUC__ >= 7
+#define OLA_FALLTHROUGH __attribute__ ((fallthrough));
+#else
+#define OLA_FALLTHROUGH
+#endif // __GNUC__ >= 7
```
Comment (peternewman) [inline include/ola/base/Macro.h]: This needs lint fixing, two spaces before the //

File: include/ola/base/Macro.h [hunk 5]
```diff
@@ -63,6 +63,30 @@
...
+#else
+#define OLA_FALLTHROUGH
+#endif // __GNUC__
```
Comment (peternewman) [inline include/ola/base/Macro.h]: Again

**Comment by yoe** [comment]
in several other places, too. I'll follow up in #1193.

File: include/ola/base/Macro.h [hunk 6]
```diff
@@ -63,6 +63,30 @@
...
+ *     case 'foo':
+ *       doFoo();
+ *       // after foo fallthrough to bar because $REASON
+ *       OLA_FALLTHROUGH
+ *     case 'bar':
+ *   @endcode
+ */
+#ifdef __GNUC__
+#if __GNUC__ >= 7
+#define OLA_FALLTHROUGH __attribute__ ((fallthrough));
```
Comment (peternewman) [inline include/ola/base/Macro.h]: Can you add a future todo/comment for why we aren't using the [[fallthrough]] variant please (as we're still on C++98).

File: plugins/usbpro/BaseRobeWidget.cpp (hunk 2 above)
Comment (peternewman) [inline plugins/usbpro/BaseRobeWidget.cpp]: Done

File: include/ola/base/Macro.h (hunk 3 above)
Comment (peternewman) [inline include/ola/base/Macro.h]: Done

File: include/ola/base/Macro.h (hunk 4 above)
Comment (peternewman) [inline include/ola/base/Macro.h]: Done

File: include/ola/base/Macro.h (hunk 5 above)
Comment (peternewman) [inline include/ola/base/Macro.h]: Done

File: include/ola/base/Macro.h (hunk 6 above)
Comment (peternewman) [inline include/ola/base/Macro.h]: Done

**Comment by daveol** [comment]
@nomis52 hey, can you review this so it can be merged?

**Comment by hepiyellow** [comment]
I am getting this error running `make`:
```
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I/opt/local/include -I./include -I./include -Wall -Wformat -W -fvisibility-inlines-hidden -D_THREAD_SAFE -I/opt/local/include -Werror -Werror -Wno-error=unused-parameter -Wno-error=deprecated-declarations -Wno-error=sign-compare -DPID_DATA_DIR=\"/usr/local/share/ola/pids\" -g -O2 -std=gnu++98 -D_THREAD_SAFE -MT common/utils/common_libolacommon_la-StringUtils.lo -MD -MP -MF common/utils/.deps/common_libolacommon_la-StringUtils.Tpo -c common/utils/StringUtils.cpp  -fno-common -DPIC -o common/utils/.libs/common_libolacommon_la-StringUtils.o
common/utils/StringUtils.cpp:420:9: error: use of undeclared identifier 'OLA_FALLTHROUGH'
        OLA_FALLTHROUGH
        ^
common/utils/StringUtils.cpp:424:9: error: use of undeclared identifier 'OLA_FALLTHROUGH'
        OLA_FALLTHROUGH
        ^
2 errors generated.
make[2]: *** [common/utils/common_libolacommon_la-StringUtils.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

My gcc:
```
$ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Running on OS X 10.11.6 (El Capitan)
and Xcode 7.3.1

Using OLA master from Git,
`commit ce879956fb035194fd1aa486e77357dce7dc61e8 `

Any ideas?

**Comment by peternewman** [comment]
Hi @hepiyellow , firstly you're better off starting a new issue, rather than commenting on a pull request that's been merged and closed for a month.

In terms of your issue, I think you'll find there's an older version of Macro.h somewhere on your machine, from a previous build/install that it's loading instead (which won't have OLA_FALLTHROUGH defined), hence why it's not working.

If this is still a problem, can you open an issue where we can continue the discussion.

**Pull request merged**